src/dune_console/progress.ml

```diff
@@ -3,25 +3,53 @@ open Stdune
 module No_flush = struct
   let status_line = ref Pp.nop
   let start () = ()
-  let status_line_len = ref 0
+  let status_line_height = ref 0
 
-  let hide_status_line () =
-    if !status_line_len > 0 then Printf.eprintf "\r%*s\r" !status_line_len ""
+  let hide_status_line =
+    (* This ansi code will clear the entire line leaving the cursor in place. *)
+    let ansi_erase = "\027[2K" in
+    (* Move the cursor up one line. *)
+    let ansi_up = "\027M" in
+    fun () ->
+      if !status_line_height > 0
+      then (
+        (* Return the cursor to the begining of the line and clear it. *)
+        Printf.eprintf "\r%s" ansi_erase;
+        for _ = 1 to !status_line_height - 1 do
+          (* If we have multiple lines, we move the cursor up and clear. *)
+          Printf.eprintf "%s%s" ansi_up ansi_erase
+        done)
   ;;
 
-  let show_status_line () = if !status_line_len > 0 then Ansi_color.prerr !status_line
+  let show_status_line () = if !status_line_height > 0 then Ansi_color.prerr !status_line
 
   let set_status_line = function
     | None ->
       hide_status_line ();
       status_line := Pp.nop;
-      status_line_len := 0
+      status_line_height := 0
     | Some line ->
       let line = Pp.map_tags line ~f:User_message.Print_config.default in
-      let line_len = String.length (Format.asprintf "%a" Pp.to_fmt line) in
       hide_status_line ();
       status_line := line;
-      status_line_len := line_len;
+      (status_line_height
+         := Format.asprintf "%a" Pp.to_fmt line
+            |> String.split_lines
+            |>
+            (* Notty_unix gives us a cross-platform way to get the terminal width. *)
+            match Notty_unix.winsize Unix.stderr with
+            | Some (width, _) ->
+              List.fold_left ~init:0 ~f:(fun line_count next_line ->
```

Comment:
All this logic looks like we're creating a problem for ourselves and then working hard to solve it. Currently, this is doing the following:

1. Pretty print the status line while wrapping long lines at a certain margin (say 80 chars)
2. Inspect the terminal settings to see if we are overflowing
3. Count the number of lines we've overflowed by.

What about just not allowing overflows in the first place.

First, if we are going to allow multiple entries in the status, let's just make that explicit by introducing:

```ocaml
val set_status : User_message.Style.t Pp.t list -> unit
```

to replace the existing `set_status_line`.

Then, whenever an entry in this status doesn't fit in one line, we just truncate it and add `...` to show it's being truncated. If the user's display is small none of this wrapping is going to look good anyway. It also makes no sense to wrap logical entries like the progress counter.